homeassistant/components/clarifai/api.py

```diff
@@ -0,0 +1,135 @@
+        request = service_pb2.ListAppsRequest()
+        response = self._stub.ListApps(request, metadata=self._metadata)
+        if _validate_response(response):
+            return {app.name: app.id for app in response.apps}
+
+    def list_app_scopes(self, app_id):
+        """List allowed scopes for configured access_token and application."""
+        request = service_pb2.MyScopesRequest(
+            user_app_id=resources_pb2.UserAppIDSet(app_id=app_id)
+        )
+        response = self._stub.MyScopes(request, metadata=self._metadata)
+        if _validate_response(response):
+            return response.scopes
+
+    def post_workflow_results(self, app_id, workflow_id, result_format, image_bytes):
+        """Run a workflow with a given image and return the results."""
+        request = service_pb2.PostWorkflowResultsRequest(
+            user_app_id=resources_pb2.UserAppIDSet(app_id=app_id),
+            workflow_id=workflow_id,
+            inputs=[
+                resources_pb2.Input(
+                    data=resources_pb2.Data(
+                        image=resources_pb2.Image(base64=image_bytes)
+                    )
+                )
+            ],
+        )
+        response = self._stub.PostWorkflowResults(request, metadata=self._metadata)
+        if _validate_response(response):
+            return _parse_workflow_response(response, result_format)
+
+
+def _validate_response(response):
+    """Validate that the response returned from a request was successful."""
+    if response.status.code != status_code_pb2.SUCCESS:
+        raise HomeAssistantError(
+            f"Request failed, status code: {str(response.status.code)}"
+        )
+    return True
+
+
+def _parse_workflow_response(response, result_format=NESTED):
+    """Parse a response containing a WorkflowResult object."""
+    result_options = [FLAT, NESTED]
+    if result_format not in result_options:
+        raise HomeAssistantError(f"Invalid workflow result format: {result_format}")
+
+    workflow_result = response.results[0]  # we send 1 input so we get 1 result
+    if result_format == NESTED:
+        return _parse_workflow_result_nested(workflow_result)
+    elif result_format == FLAT:
+        return _parse_workflow_result_flat(workflow_result)
+
+
+def _parse_workflow_result_nested(workflow_result):
+    """Parse a WorkflowResult, keeping its original nested format."""
+    return json.loads(json_format.MessageToJson(workflow_result))
+
+
+def _parse_workflow_result_flat(workflow_result):
+    """
+    Parse a WorkflowResult, flattening concepts and regions into lists.
+
+    The purpose behind flattening concepts and regions is to make them
+    more accessible in automations, which use Jinja YAML templating.
+
+    Concepts are returned as a dictionary with the concept name as the
+    key and a list of confidence values as the value. Regions are
+    returned as a list of region_info dictionaries.
+
+    Ex:
+    "concepts": {
```

Comment:
I think it would be better if you could associate each region with its respective concepts when flattening. If someone were to use the regions, it would be helpful for them to know what concepts go with that region. 
This could maybe be a list of tuples? like `(single region, [associated concepts])`, or in the data we return, you can pull the concepts out the region (from data.region.data.concepts -> data.concepts)

Reply:
Could we maintain the relationship between the concepts that are found within a region while still aggregating them and presenting them to the user at the top level?

Reply:
Another idea – maybe it would be better to instead of using flat/nested, use concepts/default? I.e. don't return the regions at all in the "flat" version, just return the aggregated dict of concepts by name and let them preserve the original structure when looking for region data? 

Also, when extracting the concepts, I just realized that there are sometimes concepts that are not returned with an individual confidence value. I think in those cases, I will exclude them from the aggregated concepts dict. 

Perhaps I'm molding this "flat" / "concepts" result format tightly to my example use case of a front door camera, but I think it's a common use case to want random access to all the concepts returned without having to crawl through the results and check each region individually for a concept. 

Maybe it's unnecessary to try to restructure the regions when the format they are returned in originally already contains a lot of contextual information?

Reply:
I think as you mentioned would be good. One list of the aggregated concepts while keeping the list of regions to be in the format they were returned in for further use.